### PR TITLE
new collaborator access widget

### LINKED ISSUE #1519: Collaborators descriptions role allowance
BE:
- [x] Update CollaboratorProps to be sections: 'all' | 'none' | Record<string ,{tableData:boolean,descriptions:boolean}>
- [x] in current data migration, update current sections to {tableData:true,descriptions:true}
- [x] Update current auth middleware to authorize collaborators to edit table data and descriptions (treat them differently if they'are not now)

FE:
- [x] User edit ui:  update current sections edit to reflect the possibility to edit tableData and descriptions
- [x] Descriptions edit: update edit allowance based on new collaborator props
- [x] Table data edit: update edit allowance based on new collaborator props

OJ: Add functionality to allow contributors to edit text sections but not change data in the reporting tables (Örjan) (TODO: REFINE access level)

### PR DESCRIPTION
**Pull request by rfontanarosa**
Part of  #1519

https://user-images.githubusercontent.com/10047945/193007965-92296fe6-3832-4d18-93c8-39724565c0b3.mov

### CODE REVIEW

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        typeof selectedSections !== 'string'
+          ? { ...selectedSections, [section]: { ...selectedSections[section], [permission]: true } }
+          : { [section]: { [permission]: true } }
+      )
+  }
+
+  const toggleOption = (section: string, permission: CollaboratorEditPropertyType): void => {
+    if (typeof selectedSections === 'string' || !selectedSections[section]?.[permission]) addOption(section, permission)
+    else if (selectedSections[section]?.[permission]) removeOption(section, permission)
+  }
```
Comment (minotogna) [inline src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx]: export this functions from hook useActions . ```import {toggleOptions...} from './hooks/useActions'```

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+              <div
+                className={classNames('fra-checkbox', {
+                  checked: selectedSections === 'none',
+                })}
+              />
+              <div className="form-field-container-label">{i18n.t(`contactPersons.none`)}</div>
+            </div>
+          </div>
+          <div className="form-container">
+            <MediaQuery maxWidth={Breakpoints.laptop - 1}>
```
Comment (minotogna) [inline src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx]: why two different components? is that really necessary ?

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  height: auto;
+}
+
+.edit-user__form-field-premissions {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  flex-direction: column;
+  padding: 12px;
+      font-size: 12px;
```
Comment (minotogna) [inline src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss]: isnt this auto formatted?

File: src/server/repository/public/user/getOne.ts [hunk 4]
```diff
@@ -28,15 +28,29 @@ export const getOne = async (
...
-    Objects.camelize
-  )
+      [value]
+    )
+    .then((data) =>
+      data
+        ? {
+            ...Objects.camelize(data),
+            roles:
+              data.roles && data.roles[0] !== null
```
Comment (minotogna) [inline src/server/repository/public/user/getOne.ts]: why first item should be not null? is this meant to be length > 0 ?

File: src/client/components/EditUserForm/EditUserForm.tsx [hunk 5]
```diff
@@ -72,6 +73,7 @@ const EditUserForm: React.FC<{ user: User }> = ({ user }) => {
     <div className="edit-user__form-container">
       <ProfilePicture userId={user.id} onChange={(profilePicture: File) => setProfilePicture(profilePicture)} />
       <TextInputFields user={user} onChange={changeUser} />
+      {user.roles[0].role === RoleName.COLLABORATOR && <CollaboratorPermissions user={user} />}
```
Comment (minotogna) [inline src/client/components/EditUserForm/EditUserForm.tsx]: i have already noticed that several times, we are using user.roles[0] to identify role for country.
however this is incorrect and inconsistent. 

1. user should have fetched all roles, not only current country role (edit user form can be access from admin page) (country role should be accessed using some utility Users.getCountryRole({user,countryIso})
2. here you need to check that countryIso exists.

do you want to fix this user.roles[0] inconsistency in this PR or a separate issue ?

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/client/components/EditUserForm/EditUserForm.tsx (hunk 5 above)
Comment (rfontanarosa) [reply src/client/components/EditUserForm/EditUserForm.tsx]: this is always true when we get users using getCountryUsers but you're right, not in that case. we can replace all the user.roles[0] with Users.getCountryRole. I can do in this pr.

Schermata 2022-09-30 alle 10 00 35

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx (hunk 2 above)
Comment (rfontanarosa) [reply src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx]: I don't know, it's the same on CountrySelectModal. this is how it appears on narrow screens. probably they are unnecessary. we can find another solution for both the modals on a different pr.

Schermata 2022-09-30 alle 09 54 15

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss (hunk 3 above)
Comment (rfontanarosa) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss]: i don't know what happened with mi ide. I save it again and it auroformatted-it.

File: src/server/repository/public/user/getOne.ts (hunk 4 above)
Comment (rfontanarosa) [reply src/server/repository/public/user/getOne.ts]: nope, this is a getOne, it doesn't return an array. I added this check because otherwise the tests fail. but I've made some changes to make it more readable.

Schermata 2022-09-30 alle 10 17 20

File: src/client/components/EditUserForm/EditUserForm.tsx (hunk 5 above)
Comment (minotogna) [reply src/client/components/EditUserForm/EditUserForm.tsx]: It's also about fetching all roles associated to each user . That is the source of the inconsistency

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx (hunk 2 above)
Comment (minotogna) [reply src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx]: Can u just fix this PR ?

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss (hunk 3 above)
Comment (minotogna) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss]: Don't you use pre commit hook ?

**Comment by minotogna** [comment]
I have also updated the description of this pr as it doesn't close the issue

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx (hunk 1 above)
Comment (rfontanarosa) [reply src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx]: moved logic and state inside useActions

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx (hunk 2 above)
Comment (rfontanarosa) [reply src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx]: ok

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss (hunk 3 above)
Comment (rfontanarosa) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss]: yes, I do, are scss included into the auto-format?

Comment (minotogna) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss]: can u check ?

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    setModalOptions({ open: true })
+  }
+
+  const _onEditPermissionsClose = () => {
+    setModalOptions({ open: false })
+  }
+
+  return (
+    <div className="edit-user__form-item edit-user__form-item-permissions">
+      <div className="edit-user__form-label">{i18n.t<string>('Permissions')}</div>
```
Comment (minotogna) [inline src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx]: what's this??

**Review by minotogna — CHANGES_REQUESTED** [review]
- fix user roles[0]
- fix i18n

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx (hunk 6 above)
Comment (rfontanarosa) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx]: Schermata 2022-09-30 alle 12 44 32

Comment (minotogna) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx]: I meant using 'Permissions' is not consistent on how we write label key in the js translation files. fix it

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx (hunk 6 above)
Comment (rfontanarosa) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx]: fixed

File: src/client/components/EditUserForm/EditUserForm.tsx [hunk 7]
```diff
@@ -68,10 +69,13 @@ const EditUserForm: React.FC<{ user: User }> = ({ user }) => {
 
   if (!user) return null
 
+  const userRole = Users.getCountryRole(user, countryIso)
+
   return (
     <div className="edit-user__form-container">
       <ProfilePicture userId={user.id} onChange={(profilePicture: File) => setProfilePicture(profilePicture)} />
       <TextInputFields user={user} onChange={changeUser} />
+      {userRole.role === RoleName.COLLABORATOR && <CollaboratorPermissions userRole={userRole} />}
```
Comment (minotogna) [inline src/client/components/EditUserForm/EditUserForm.tsx]: again, you must check that there is countryIso

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/client/components/EditUserForm/EditUserForm.tsx (hunk 7 above)
Comment (rfontanarosa) [reply src/client/components/EditUserForm/EditUserForm.tsx]: it does it on line 72, but it doesn't check it the role is undefined. fixed.

**Review by minotogna — APPROVED** [review]

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  const { headerLabel, onClose, open, userRole } = props
+
+  const dispatch = useAppDispatch()
+  const i18n = useTranslation()
+  const assessmentSections = useAssessmentSections()
+
+  const options = assessmentSections
+    .reduce((prev, curr): Array<SubSection> => [...prev, ...curr.subSections], [])
+    .filter((subSection: SubSection) => subSection.props.anchor)
+    .reduce((prev, curr): Record<string, string> => ({ ...prev, [curr.uuid]: curr.props.anchor }), {})
```
Comment (sorja) [inline src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx]: Can we avoid triple loop?

File: src/server/repository/public/user/getOne.ts [hunk 9]
```diff
@@ -28,15 +28,28 @@ export const getOne = async (
...
-    Objects.camelize
-  )
+      [value]
+    )
+    .then((data) => {
+      if (!data) return undefined
+      return {
+        ...Objects.camelize(data),
+        roles:
+          data.roles && data.roles[0] !== null
```
Comment (sorja) [inline src/server/repository/public/user/getOne.ts]: `data?.roles?.[0]` no need double check data.roles

File: src/i18n/resources/en.js [hunk 10]
```diff
@@ -524,7 +525,7 @@ The FRA team
     confirmDelete: 'Remove {{user}}?',
     allUsers: 'All collaborators',
     info: 'Info',
-    tableAccess: 'Table access',
+    permissions: 'Permissions',
```
Comment (sorja) [inline src/i18n/resources/en.js:528]: was it added to i18n issue?

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    .reduce((prev, curr): Array<SubSection> => [...prev, ...curr.subSections], [])
+    .filter((subSection: SubSection) => subSection.props.anchor)
+    .reduce((prev, curr): Record<string, string> => ({ ...prev, [curr.uuid]: curr.props.anchor }), {})
+
+  const properties = (userRole.props as CollaboratorProps) || undefined
+  const sections = Objects.isEmpty(properties) ? 'none' : properties.sections
+
+  const tableDataPermissions = Object.entries(sections)
+    .filter(([_, v]) => v.tableData === true)
+    .map(([k, _]) => options[k])
```
Comment (sorja) [inline src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.tsx]: map and filter should be combined to avoid double loop

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  height: auto;
+}
+
+.edit-user__form-field-premissions {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  flex-direction: column;
+  padding: 12px;
+  font-size: 12px;
```
Comment (sorja) [inline src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss:13]: do we have config for these? `spacing-x`?

File: src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                      key={`${section}-${permission}`}
+                      className="form-field-selector"
+                      onClick={() => toggleOption(section, permission)}
+                      onMouseDown={(e) => e.stopPropagation()}
+                      aria-hidden="true"
+                    >
+                      <div
+                        className={classNames('fra-checkbox', {
+                          checked:
+                            typeof selectedSections !== 'string' &&
```
Comment (sorja) [inline src/client/components/CollaboratorAccessModal/CollaboratorAccessModal.tsx:126]: should we avoid nesting so much, a bit hard to follow the code

File: src/client/components/CollaboratorAccessModal/hooks/useActions.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+) => {
+  const [selectedSections, setSelectedSections] = useState(sections)
+
+  const toggleOptions = (permission: CollaboratorEditPropertyType): void => {
+    const enabled =
+      typeof selectedSections !== 'string'
+        ? Object.entries(selectedSections).filter(([_, section]) => section[permission] === true).length <
+          Object.keys(options).length
+        : true
+    const newSelectedSections = typeof selectedSections !== 'string' ? Objects.cloneDeep(selectedSections) : {}
```
Comment (sorja) [inline src/client/components/CollaboratorAccessModal/hooks/useActions.ts:17]: should we add helper for `typeof selectedSections !== 'string'`? used 9 times

Comment (minotogna) [reply src/client/components/CollaboratorAccessModal/hooks/useActions.ts:17]: or event just common variables. like isAll, isNone ?

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss (hunk 12 above)
Comment (minotogna) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss:13]: we do, but not sure why they sometimes get ignored

Comment (rfontanarosa) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss:13]: we don't have a 12px one

Schermata 2022-10-03 alle 10 31 13

File: src/client/components/CollaboratorAccessModal/hooks/useActions.ts (hunk 14 above)
Comment (rfontanarosa) [reply src/client/components/CollaboratorAccessModal/hooks/useActions.ts:17]: I'm having a lot of issues from typescript trying to wrap that condition somewhere else. let's skip it for now, I can show you what happens later.

File: src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss (hunk 12 above)
Comment (rfontanarosa) [reply src/client/components/EditUserForm/CollaboratorPermissions/CollaboratorPermissions.scss:13]: let's open another issue about it because it affects the whole user form

**Review by sorja — APPROVED** [review]
